src/aotcompile.cpp

```diff
@@ -248,6 +248,32 @@ static void makeSafeName(GlobalObject &G)
         G.setName(StringRef(SafeName.data(), SafeName.size()));
 }
 
+static void jl_ci_cache_lookup(const jl_cgparams_t &cgparams, jl_method_instance_t *mi, size_t world, jl_code_instance_t **ci_out, jl_code_info_t **src_out)
+{
+    jl_value_t *ci = cgparams.lookup(mi, world, world);
+    JL_GC_PROMISE_ROOTED(ci);
+    jl_code_instance_t *codeinst = NULL;
+    if (ci != jl_nothing) {
+        codeinst = (jl_code_instance_t*)ci;
+        *src_out = (jl_code_info_t*)codeinst->inferred;
+        jl_method_t *def = codeinst->def->def.method;
+        if ((jl_value_t*)*src_out == jl_nothing)
```

Comment:
What with `rettype_const` methods, those don't have a `ci.inferred`?

Reply:
Also, what happens if the JIT deletes an inferred code instance, e.g. https://github.com/JuliaLang/julia/blob/8b010b5f1b33c0397f4019902163cc013ea59a7f/src/codegen.cpp#L6597 from https://github.com/JuliaLang/julia/pull/35655, will that not mutate the external cache then?

Reply:
> What with rettype_const methods, those don't have a ci.inferred?

Same as before, they won't get an LLVM function emitted for them.

Reply:
> Also, what happens if the JIT deletes an inferred code instance,

Fair point, that should probably be turned off.

Reply:
> Same as before, they won't get an LLVM function emitted for them.

We currently do, though. If you call `jl_create_native` for `kernel() = return` you get the following un-optimized IR:
```llvm
define internal void @julia_kernel_39() !dbg !5 {
top:
  %0 = call %jl_value_t*** @julia.ptls_states()
  %1 = bitcast %jl_value_t*** %0 to %jl_value_t addrspace(10)**
  %2 = getelementptr inbounds %jl_value_t addrspace(10)*, %jl_value_t addrspace(10)** %1, i64 4
  %3 = bitcast %jl_value_t addrspace(10)** %2 to i64**
  %4 = load i64*, i64** %3, !tbaa !7, !invariant.load !4
  ret void, !dbg !11
}
```

It's only when calling this function from another that the return type gets folded.

Doing the same with the proposed PR and a `lookup` cgparam yields `Refusing to automatically run type inference with custom cache lookup`.

Reply:
Ah, that just seems like a bug then. We should just allow that case.

Reply:
Ok, I see what's happening here. In the regular case it actually does call type inference then, which will then give it a CodeInfo, even though we never cache those because they're not actually supposed to be used. I would say for now, let's just filter these in the AbstractInterpreter implementation that calls into here and maybe in the future, we can have it synthesize an appropriate body, without forcing it re-do type inference.